proposals/0111-linear-types.rst

```diff
@@ -907,6 +909,159 @@ or ``case`` in the surface syntax), we want to infer the multiplicity
+(including a multiplicity variable):
+
+::
+
+   let
+     %p1 pat1 = rhs1
+     …
+     %pn patn = rhsn
+   in body
+
+if
+
+* The binding is not top-level
+* The binding is non-recursive
+* The binding is a pattern binding (including a simple variable)
+  ``p=e``
+* Either ``p`` is of the form ``!p'`` or ``p`` is a variable. In
+  particular neither ``x@y`` nor ``(x)`` are covered by “is a
+  variable”
+
+If there is a multiplicity annotation, the binding is not
+generalised. So ``let %m x = e in b`` and ``(\(%m x) ->  b) e`` have
+the same typing rule.
+
+When the multiplicity annotation isn't specified, the multiplicity is
+inferred, so that both
+
+::
+
+   \(%1 z) -> let !(x, y) = z in (y, x)
+   let !(x, y) = u in (x, x)
+
+are well typed. Recursive bindings, toplevel bindings and non-variable
+lazy-pattern bindings are always inferred to have multiplicity
+``Many``.
+
+Note that, in particular, function bindings are not allowed:
+
+::
+
+   -- Doesn't work
+   let %p f x y = rhs in body
+
+instead write
+
+::
+
+   -- pat can be a variable
+   let %p f = \x y -> rhs in body
+
+When typechecking ``let %p pat = rhs in body`` the ``rhs`` is consumed
+with multiplicity ``p`` and the variables of ``pat`` must be consumed
+in ``body`` with multiplicity ``p`` (if ``pat`` has some non-linear
+fields, then the variables are scaled appropriately as per
+`Constructors & pattern-matching`_). The pattern ``pat`` can be an
+arbitrary, nested, pattern.
+
+Here are a few examples that illustrate the typing rules
+
+::
+
+   -- good
+   let %1 x = u in x
+
+   -- bad
+   let %1 x = u in (x, x)
+
+   -- bad
+   let %1 x = u in let %Many y = x in …
+
+   -- good
+   let %1 (x, y) = u in (y, x)
+
+   -- bad
+   let %1 (x, y) = u in x
+
+   -- good
+   let %1 (Ur x) = u in (x, x, x)
+
+Non-variable let-bound patterns must be strict
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Non-variable linear patterns can't be lazy (see `Lazy patterns`_). As
+a consequence, non-variable let-bound patterns must be annotated with
+a ``!`` (because let-bound patterns are lazy by default, as opposed to
+case-bound patterns which are strict by default).
+
+::
+
+   -- good
+   let %1 x = u in …
+
+   -- good
+   let %1 !x = u in …
+
+   -- bad
+   let %1 (x, y) = u in …
+
+   -- good
+   let %1 !(x, y) = u in …
+
+   -- inferred unrestricted
+   let (x, y) = u in …
+
+Non-variable linear patterns are monomorphic
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. _`Non-variable linear patterns are monomorphic`
+
+Variables in non-variable multiplicity-annotated let-patterns are bound to
+monomorphic types (see `Let bindings and polymorphism`_ for the
+reasoning and a discussion). Unannotated non-variable let-patterns are
+inferred to be unrestricted (by default, since ``-XLinearTypes``
+implies ``-XNoMonoLocalBinds``, only toplevel bindings, which are
+always unrestricted anyway, are inferred to be polymorphic).
```

Comment:
I admit to having been a little careless in my phrasing here. Unless I'm missing something, this is all fine. Case (1) is the only one that does generalisation which, as you notice below, is the real problem. So in case 1 the binding needs to be unrestricted. But since it doesn't capture any variable, unrestricted is the most general type anyway. So we don't lose any generality.

Any suggestion for better phrasing?